### PR TITLE
Projects sync ignore excluded dirs

### PR DESCRIPTION
**Pull request by stackoverflow**

### CODE REVIEW

File: src/main/kotlin/org/pkl/intellij/packages/PklProjectService.kt [hunk 1]
```diff
@@ -304,15 +305,24 @@ class PklProjectService(private val project: Project) :
   }
 
   private fun discoverProjectFiles(): List<VirtualFile> {
+    val excludedRoots = mutableSetOf<Path>()
```
Comment (bioball) [inline src/main/kotlin/org/pkl/intellij/packages/PklProjectService.kt:308]:
```suggestion
    val excludedRoots = mutableSetOf<VirtualFile>()
```

File: src/main/kotlin/org/pkl/intellij/packages/PklProjectService.kt [hunk 2]
```diff
@@ -304,15 +305,24 @@ class PklProjectService(private val project: Project) :
...
     return project.modules
       .toList()
-      .flatMap { ModuleRootManager.getInstance(it).contentRoots.toList() }
+      .flatMap { mod ->
+        val rootManager = ModuleRootManager.getInstance(mod)
+        excludedRoots.addAll(rootManager.excludeRoots.map { it.toNioPath() })
```
Comment (bioball) [inline src/main/kotlin/org/pkl/intellij/packages/PklProjectService.kt:313]:
```suggestion
        excludedRoots.addAll(rootManager.excludeRoots)
```

File: src/main/kotlin/org/pkl/intellij/packages/PklProjectService.kt [hunk 3]
```diff
@@ -304,15 +305,24 @@ class PklProjectService(private val project: Project) :
...
       .flatMap { contentRoot ->
         // no need for async here because we are already inside a background task.
         contentRoot.refresh(false, true)
         buildList {
           VfsUtil.processFileRecursivelyWithoutIgnored(contentRoot) { file ->
-            if (file.fileType == PklFileType && file.name == PKL_PROJECT_FILENAME) {
+            if (
+              file.fileType == PklFileType &&
+                file.name == PKL_PROJECT_FILENAME &&
+                !excludedRoots.any { it.isAncestor(file.toNioPath()) }
```
Comment (bioball) [inline src/main/kotlin/org/pkl/intellij/packages/PklProjectService.kt:324]:
```suggestion
                !excludedRoots.any { VfsUtilCore.isAncestor(it, file, true) }
```

**Review by bioball — CHANGES_REQUESTED** [review]
Calling `.toNioPath` can possibly throw if the file isn't backed by a `Path`. But, thankfully, there's a handy utility method that we can use from the IJ SDK.

BTW: I think we should also add some logic to `org.pkl.intellij.notification.PklSyncProjectNotificationProvider` to not show the "Sync Project" banner if it resides inside an excluded root. Maybe the logic can be added to `org.pkl.intellij.psi.PklModule#isInPklProject` to return `false` if it's inside an excluded root?

File: src/main/kotlin/org/pkl/intellij/notification/PklSyncProjectNotificationProvider.kt [hunk 4]
```diff
@@ -88,4 +92,10 @@ class PklSyncProjectNotificationProvider(project: Project) : EditorNotificationP
       null
     }
   }
+
+  private fun isExcluded(project: Project, file: VirtualFile): Boolean {
+    val excludes =
+      project.modules.flatMap { ModuleRootManager.getInstance(it).excludeRoots.toList() }
+    return excludes.any { VfsUtilCore.isAncestor(it, file, true) }
+  }
```
Comment (bioball) [inline src/main/kotlin/org/pkl/intellij/notification/PklSyncProjectNotificationProvider.kt]:
```suggestion
    for (module in project.modules) {
      for (root in module.rootManager.excludeRoots) {
        if (VfsUtilCore.isAncestor(root, file, true)) return true
      }
    }
    return false
```

File: src/main/kotlin/org/pkl/intellij/psi/PklModuleImpl.kt [hunk 5]
```diff
@@ -254,7 +254,13 @@ class PklModuleImpl(viewProvider: FileViewProvider) :
...
 
   override val isInPklProject: Boolean
-    get() = pklProjectDir != null
+    get() = pklProjectDir != null && !isExcluded(project, virtualFile)
+
+  private fun isExcluded(project: Project, file: VirtualFile): Boolean {
+    val excludes =
+      project.modules.flatMap { ModuleRootManager.getInstance(it).excludeRoots.toList() }
+    return excludes.any { VfsUtilCore.isAncestor(it, file, true) }
+  }
```
Comment (bioball) [inline src/main/kotlin/org/pkl/intellij/psi/PklModuleImpl.kt:263]: I think this is too expensive in its current form :(

We either need to find a way to make this cheap (maybe cache it, with correct cache invalidation), or let's just move it back to `PklSyncNotificationProvider`.

Comment (stackoverflow) [reply src/main/kotlin/org/pkl/intellij/psi/PklModuleImpl.kt:263]: This is executed once each time you open or switch to a new file. I can't see this being too expensive unless you have thousands of excluded folders.

Comment (bioball) [reply src/main/kotlin/org/pkl/intellij/psi/PklModuleImpl.kt:266]: Ah, right, we currently only call this inside `PklSyncProjectNotificationProvider`. I had thought that we use this method on every resolve.

**Review by bioball — APPROVED** [review]

**Pull request merged**